src/System.Net.NameResolution/src/System/Net/NameResolutionPal.Windows.cs

```diff
@@ -385,10 +387,288 @@ public static void EnsureSocketsAreInitialized()
+                    var socketAddress = new ReadOnlySpan<byte>(result->ai_addr, result->ai_addrlen);
+
+                    if (result->ai_family == AddressFamily.InterNetwork)
+                    {
+                        if (socketAddress.Length == SocketAddressPal.IPv4AddressSize)
+                            ipAddress = CreateIPv4Address(socketAddress);
+                    }
+                    else if (SocketProtocolSupportPal.OSSupportsIPv6 && result->ai_family == AddressFamily.InterNetworkV6)
+                    {
+                        if (socketAddress.Length == SocketAddressPal.IPv6AddressSize)
+                            ipAddress = CreateIPv6Address(socketAddress);
+                    }
+
+                    if (ipAddress != null)
+                        addresses.Add(ipAddress);
+
+                    result = result->ai_next;
+                }
+
+                if (canonicalName == null)
+                    canonicalName = state.HostAddress;
+
+                state.CompleteAsyncResult(new IPHostEntry
+                {
+                    HostName = canonicalName,
+                    Aliases = Array.Empty<string>(),
+                    AddressList = addresses.ToArray()
+                });
+            }
+            finally
+            {
+                state.Dispose();
+            }
+        }
+
+        private static unsafe IPAddress CreateIPv4Address(ReadOnlySpan<byte> socketAddress)
+        {
+            long address = (long)SocketAddressPal.GetIPv4Address(socketAddress) & 0x0FFFFFFFF;
+            return new IPAddress(address);
+        }
+
+        private static unsafe IPAddress CreateIPv6Address(ReadOnlySpan<byte> socketAddress)
+        {
+            Span<byte> address = stackalloc byte[IPAddressParserStatics.IPv6AddressBytes];
+            uint scope;
+            SocketAddressPal.GetIPv6Address(socketAddress, address, out scope);
+
+            return new IPAddress(address, (long)scope);
+        }
+
+        #region GetAddrInfoAsync Helper Classes
+
+        private sealed unsafe class GetAddrInfoExState : IDisposable
+        {
+            private IntPtr _context;
+            private int _callbackStartedOrCanceled;
+            private DnsResolveAsyncResult _asyncResult;
+
+            public string HostAddress { get; }
+
+            public GetAddrInfoExState(string hostAddress, DnsResolveAsyncResult asyncResult)
+            {
+                HostAddress = hostAddress;
+                _asyncResult = asyncResult;
+            }
+
+            public void SetQueryContext(IntPtr context)
+            {
+                _context = context;
+            }
+
+            public bool SetCallbackStartedOrCanceled()
+            {
+                return Interlocked.Exchange(ref _callbackStartedOrCanceled, 1) == 0;
+            }
+
+            public void CompleteAsyncResult(object o)
+            {
+                // We don't want to expose the GetAddrInfoEx callback thread to user code.
```

Comment:
Is this just one of the I/O completion port threads?  We expose it in other places, like sockets.  We can leave it this way for now, with the queuing, but we might want to revisit whether this is really necessary.
cc: @geoffkizer

Reply:
This is a native windows threadpool thread.

Reply:
Ah. Ok.